api/node/typescript/models.ts

```diff
@@ -0,0 +1,436 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+import * as napi from "../rust-module.cjs";
+
+class ModelIterator<T> implements Iterator<T> {
+    private row: number;
+    private model: Model<T>;
+
+    constructor(model: Model<T>) {
+        this.model = model;
+        this.row = 0;
+    }
+
+    public next(): IteratorResult<T> {
+        if (this.row < this.model.rowCount()) {
+            const row = this.row;
+            this.row++;
+            return {
+                done: false,
+                value: this.model.rowData(row),
+            };
+        }
+        return {
+            done: true,
+            value: undefined,
+        };
+    }
+}
+
+/**
+ * Model<T> is the interface for feeding dynamic data into
+ * `.slint` views.
+ *
+ * A model is organized like a table with rows of data. The
+ * fields of the data type T behave like columns.
+ *
+ * @template T the type of the model's items.
+ *
+ * ### Example
+ * As an example let's see the implementation of {@link ArrayModel}
+ *
+ * ```js
+ * export class ArrayModel<T> extends Model<T> {
+ *    private a: Array<T>
+ *
+ *   constructor(arr: Array<T>) {
+ *        super();
+ *        this.a = arr;
+ *    }
+ *
+ *    rowCount() {
+ *        return this.a.length;
+ *    }
+ *
+ *    rowData(row: number) {
+ *       return this.a[row];
+ *    }
+ *
+ *    setRowData(row: number, data: T) {
+ *        this.a[row] = data;
+ *        this.notifyRowDataChanged(row);
+ *    }
+ *
+ *    push(...values: T[]) {
+ *        let size = this.a.length;
+ *        Array.prototype.push.apply(this.a, values);
+ *        this.notifyRowAdded(size, arguments.length);
+ *    }
+ *
+ *    remove(index: number, size: number) {
+ *        let r = this.a.splice(index, size);
+ *        this.notifyRowRemoved(index, size);
+ *    }
+ *
+ *    get length(): number {
+ *        return this.a.length;
+ *    }
+ *
+ *    values(): IterableIterator<T> {
+ *        return this.a.values();
+ *    }
+ *
+ *    entries(): IterableIterator<[number, T]> {
+ *        return this.a.entries()
+ *    }
+ *}
+ * ```
+ */
+export abstract class Model<T> implements Iterable<T> {
+    /**
+     * @hidden
+     */
+    modelNotify: napi.ExternalObject<napi.SharedModelNotify>;
+
+    constructor(modelNotify?: napi.ExternalObject<napi.SharedModelNotify>) {
+        if (modelNotify === undefined) {
+            this.modelNotify = napi.jsModelNotifyNew(this);
+            return;
+        }
+
+        this.modelNotify = modelNotify;
```

Comment:
The early return can be dangerous when adding new code at the bottom. I suggest to handle the null case in one line:

```suggestion
        this.modelNotify = modelNotify ?? napi.jsModelNotifyNew(this);
```